mmdet3d/models/dense_heads/primitive_head.py

```diff
@@ -0,0 +1,816 @@
+                    point_boundary_offset_xy[sel_global] = center - x[sel]
+            # Get the boundary points here
+            alldist = torch.abs(
+                torch.sum(x * plane_back[:3], 1) + plane_back[-1])
+            mind = alldist.min()
+            sel = torch.abs(alldist - mind) < self.train_cfg['dist_thresh']
+            if self.primitive_mode == 'xy':
+                if torch.sum(sel) > self.train_cfg['num_point'] and torch.var(
+                        alldist[sel]) < self.train_cfg['var_thresh']:
+                    center = torch.as_tensor([
+                        torch.mean(x[sel][:, 0]),
+                        torch.mean(x[sel][:, 1]), (zmin + zmax) / 2.0
+                    ]).to(device)
+                    sel_global = ind[sel]
+                    point_boundary_mask_xy[sel_global] = 1.0
+                    point_boundary_sem_xy[sel_global] = torch.as_tensor([
+                        center[0], center[1], center[2], zmax - zmin,
+                        (pts_semantic_mask[ind][0])
+                    ]).to(device)
+                    point_boundary_offset_xy[sel_global] = center - x[sel]
+
+        if self.primitive_mode == 'z':
+            return (point_boundary_mask_z, point_boundary_sem_z,
+                    point_boundary_offset_z)
+        elif self.primitive_mode == 'xy':
+            return (point_boundary_mask_xy, point_boundary_sem_xy,
+                    point_boundary_offset_xy)
+        elif self.primitive_mode == 'line':
+            return (point_line_mask, point_line_sem, point_line_offset)
+        else:
+            NotImplementedError
+
+    def primitive_decode_scores(self, net, end_points, num_class, mode=''):
+        net_transposed = net.transpose(2, 1)  # (batch_size, 1024, ..)
+
+        base_xyz = end_points['aggregated_points_' +
+                              mode]  # (batch_size, num_proposal, 3)
+        center = base_xyz + net_transposed[:, :, 0:
+                                           3]  # (batch_size, num_proposal, 3)
+        end_points['center_' + mode] = center
+
+        if mode in ['z', 'xy']:
+            end_points['size_residuals_' + mode] = net_transposed[:, :, 3:3 +
+                                                                  self.num_dim]
+
+        end_points['sem_cls_scores_' + mode] = net_transposed[:, :, 3 +
+                                                              self.num_dim:]
+
+        return center, end_points
+
+    def check_upright(self, para_points):
+        return (para_points[0][-1] == para_points[1][-1]) and (
+            para_points[1][-1]
+            == para_points[2][-1]) and (para_points[2][-1]
+                                        == para_points[3][-1])
+
+    def check_z(self, plane_equ, para_points):
+        return torch.sum(para_points[:, 2] +
+                         plane_equ[-1]) / 4.0 < self.train_cfg['lower_thresh']
+
+    def get_linesel(self, points, xmin, xmax, ymin, ymax):
```

Comment:
`get_linesel` and `get_linesel2` are not good names. `get_lower_linesel` and `get_upper_linesel` might be better.